core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -309,8 +543,8 @@ protected void responseWithFailure(Throwable t) {
 
         @Override
         protected void doRun() throws Exception {
-            try (Releasable shardReference = getIndexShardOperationsCounter(request.internalShardId)) {
-                shardOperationOnReplica(request.internalShardId, request);
+            try (Releasable shardReference = getIndexShardOperationsCounter(request.internalShardRouting.shardId())) {
```

Comment:
This breaks the `ShardReplicationTests.testReplicasCounter()` at least it hangs now. This is not a problem with this pr but more with the test - it needs the internalShardRouting now to be initialized, else we get an NPE in the test.